include/circt/Dialect/HW/HWOpInterfaces.h

```diff
@@ -45,25 +45,70 @@ struct PortInfo : public ModulePort {
-      if (port.isOutput())
-        outputs.push_back(port);
-      else
-        inputs.push_back(port);
-    }
+  explicit ModulePortInfo(ArrayRef<PortInfo> inputs,
+                          ArrayRef<PortInfo> outputs) {
+    ports.insert(ports.end(), inputs.begin(), inputs.end());
+    numInputs = ports.size();
+    ports.insert(ports.end(), outputs.begin(), outputs.end());
   }
 
-  /// This contains a list of the input and inout ports.
-  SmallVector<PortInfo> inputs;
-  /// This is a list of the output ports.
-  SmallVector<PortInfo> outputs;
+  explicit ModulePortInfo(ArrayRef<PortInfo> mergedPorts)
+      : ports(mergedPorts.begin(), mergedPorts.end()), numInputs(0) {
+    std::stable_sort(ports.begin(), ports.end(),
+                     [](const PortInfo &lhs, const PortInfo &rhs) {
+                       return !lhs.isOutput() && rhs.isOutput();
+                     });
+    while (numInputs < ports.size() && !ports[numInputs].isOutput())
+      ++numInputs;
+  }
+
+  using iterator = SmallVector<PortInfo>::iterator;
+  using const_iterator = SmallVector<PortInfo>::const_iterator;
+  iterator begin_input() { return ports.begin(); }
+  iterator end_input() { return ports.begin() + numInputs; }
+  const_iterator begin_input() const { return ports.begin(); }
+  const_iterator end_input() const { return ports.begin() + numInputs; }
+  iterator begin_output() { return end_input(); }
+  iterator end_output() { return ports.end(); }
+  const_iterator begin_output() const { return end_input(); }
+  const_iterator end_output() const { return ports.end(); }
+  iterator begin() { return ports.begin(); }
+  iterator end() { return ports.end(); }
+  const_iterator begin() const { return ports.begin(); }
+  const_iterator end() const { return ports.end(); }
+
+  llvm::iterator_range<iterator> all() { return {begin(), end()}; }
+  llvm::iterator_range<const_iterator> all() const { return {begin(), end()}; }
+  llvm::iterator_range<iterator> inputs() {
+    return {begin_input(), end_input()};
+  }
+  llvm::iterator_range<const_iterator> inputs() const {
+    return {begin_input(), end_input()};
+  }
+  llvm::iterator_range<iterator> outputs() {
+    return {begin_output(), end_output()};
+  }
+  llvm::iterator_range<const_iterator> outputs() const {
+    return {begin_output(), end_output()};
+  }
+
+  size_t size() const { return ports.size(); }
+  size_t size_inputs() const { return numInputs; }
+  size_t size_outputs() const { return ports.size() - numInputs; }
+
+  PortInfo &at(size_t idx) { return ports[idx]; }
+  PortInfo &at_input(size_t idx) { return ports[idx]; }
+  PortInfo &at_output(size_t idx) { return ports[idx + numInputs]; }
+
+  const PortInfo &at(size_t idx) const { return ports[idx]; }
+  const PortInfo &at_input(size_t idx) const { return ports[idx]; }
+  const PortInfo &at_output(size_t idx) const { return ports[idx + numInputs]; }
+
+  /// This contains a list of all ports.  Input first.
+  SmallVector<PortInfo> ports;
```

Comment:
Why are the member variables public? Since they must be kept in sync, it's probably better to keep them private.

Reply:
there were a few places that were modifying things directly, but I've issolated those.

Reply:
I'd be supportive of a completely immutable ModulePortInfo and requiring anything doing mutation to have a local mutated list of ports and recreate the ModulePortInfo from that.